### PR TITLE
Surface the number of threads used by ctx->runner() for ParallelMapIteratorBase

### PR DESCRIPTION
**Pull request by feihugis**
This PR surfaces the number of threads used by `ctx->runner()` for [ParallelMapIteratorBase::AddTunableParameter()](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/parallel_map_iterator.cc#L63-L66).

### CODE REVIEW

**Comment by feihugis** [comment]
@jsimsa Could you help review this PR when you have time?

**Comment by jsimsa** [comment]
Hi @feihugis thank you for the initiative. Unfortunately, this is not the right way to address the TODO.

The problem with your solution is that it assumes that the threadpool use for executing parallel map is always the interop threadpool, which is not the case. The input pipeline can be configured with a custom threadpool (see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/ops/threadpool.py and https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/experimental/threadpool_dataset_op.cc) and also using a single-threaded executor (when the `use_inter_op_parallelism attribute of the ParallelMapDataset is set to false).

I think that the proper way to address the TODO is to introduce a new member to the `IteratorContext` called `runner_threadpool_size`, which is initialized with `ctx->device()->tensorflow_cpu_worker_threads()` when created and this value is overridden when appropriate (e.g. when `override_threadpool` or `use_inter_op_parallelism=False` is used). @mrry does that make sense?

**Comment by jsimsa** [comment]
Talked to @mrry offline and updated my original comment accordingly.

**Comment by feihugis** [comment]
@jsimsa and @mrry Thanks for your comments and suggestions! Yes, the current implementation did not consider some cases. Will change the code as your suggestion.

**Comment by feihugis** [comment]
@jsimsa `runner_threadpool_size` has been added to `IteratorContext`. Could you have a look at these changes?

File: tensorflow/core/framework/dataset.h [hunk 1]
```diff
@@ -278,6 +278,9 @@ class IteratorContext {
     // Function call support.
     std::function<void(std::function<void()>)> runner = nullptr;
 
+    // Number of threads for the runner
```
Comment (jsimsa) [inline tensorflow/core/framework/dataset.h]: Number of threads used for executing user-defined functions.

**Review by jsimsa — DISMISSED** [review]
Thanks. One minor comment, otherwise looks good to me.

**Comment by feihugis** [comment]
@jsimsa Thanks for your review. The comment has been revised. The failures in the last test seem to be unrelated to this PR.

**Review by jsimsa — DISMISSED** [review]

**Comment by feihugis** [comment]
@jsimsa The code has been revised to solve the conflicts caused by the recent commits. Could you review these changes at [here](https://github.com/tensorflow/tensorflow/pull/23233/files#diff-696c1a7a179fff088f071807abd4f81aR67) and [here](https://github.com/tensorflow/tensorflow/pull/23233/files#diff-11a06473028adea81ee960e8b522a0fcR507)?

**Review by jsimsa — DISMISSED** [review]

**Comment by jsimsa** [comment]
@feihugis your PR looks good but it prompted me to perform an internal refactoring ... please sync your branch and resolve conflicts ... thanks!

**Comment by feihugis** [comment]
@jsimsa The refactoring makes the code clean! The conflicts have been resolved. Could you review the change?

File: tensorflow/core/kernels/data/iterator_ops.cc [hunk 2]
```diff
@@ -145,6 +145,8 @@ class IteratorResource : public ResourceBase {
       params.allocator_getter = [device](AllocatorAttributes attrs) {
         return device->GetAllocator(attrs);
       };
+      params.runner_threadpool_size =
```
Comment (jsimsa) [inline tensorflow/core/kernels/data/iterator_ops.cc]: Is this still needed? The `Params(OpKernelContext*)` constructors should take care of that.

File: tensorflow/core/kernels/data/iterator_ops.cc [hunk 3]
```diff
@@ -1007,6 +1009,7 @@ void IteratorGetNextSyncOp::Compute(OpKernelContext* ctx) {
   bool end_of_sequence = false;
   IteratorContext::Params params(ctx);
   params.function_library = iterator->function_library();
+
```
Comment (jsimsa) [inline tensorflow/core/kernels/data/iterator_ops.cc]: undo

File: tensorflow/core/kernels/data/iterator_ops.cc [hunk 4]
```diff
@@ -1209,7 +1212,6 @@ class DeserializeIteratorOp : public OpKernel {
   }
 };
 
-
```
Comment (jsimsa) [inline tensorflow/core/kernels/data/iterator_ops.cc]: undo

File: tensorflow/core/framework/dataset.h [hunk 5]
```diff
@@ -311,6 +314,9 @@ class IteratorContext {
     // Function call support.
     std::function<void(std::function<void()>)> runner = nullptr;
 
+    // Number of threads used for executing user-defined functions.
```
Comment (jsimsa) [inline tensorflow/core/framework/dataset.h]: let's keep the members sorted alphabetically

Comment (jsimsa) [reply tensorflow/core/framework/dataset.h]: ignore this comment

File: tensorflow/core/kernels/data/iterator_ops.cc (hunk 3 above)
Comment (feihugis) [reply tensorflow/core/kernels/data/iterator_ops.cc]: Done

File: tensorflow/core/kernels/data/iterator_ops.cc (hunk 4 above)
Comment (feihugis) [reply tensorflow/core/kernels/data/iterator_ops.cc]: Done

File: tensorflow/core/framework/dataset.h (hunk 5 above)
Comment (feihugis) [reply tensorflow/core/framework/dataset.h]: These members are in the alphabetical order, right?

**Comment by feihugis** [comment]
@jsimsa Thanks for your comments! The code has been revised. The oder of the members for `IteratorContext::Params` is as below. Does it look good?

```
    // The Allocator to be used to allocate the output of an iterator.
    std::function<Allocator*(AllocatorAttributes)> allocator_getter = nullptr;

    // Interface to operating system functionality.
    Env* env = nullptr;

    // The FunctionLibraryDefinition used to look up user-defined functions.
    std::shared_ptr<const FunctionLibraryDefinition> function_library = nullptr;

    // The FunctionLibraryRuntime object to be used to make function calls.
    FunctionLibraryRuntime* lib = nullptr;

    // If non-null, identifies the object used for performance modeling.
    std::shared_ptr<model::Model> model = nullptr;

    // Function call support.
    std::function<void(std::function<void()>)> runner = nullptr;

    // Number of threads used for executing user-defined functions.
    int32 runner_threadpool_size = 0;

    // The `StatsAggregator` object to record statistics about the iterator.
    std::shared_ptr<StatsAggregator> stats_aggregator = nullptr;
```

File: tensorflow/core/kernels/data/iterator_ops.cc (hunk 2 above)
Comment (feihugis) [reply tensorflow/core/kernels/data/iterator_ops.cc]: Oh, missed this one. Have removed it now.

File: tensorflow/core/kernels/data/parallel_map_iterator.cc [hunk 6]
```diff
@@ -65,9 +65,7 @@ class ParallelMapIterator : public DatasetBaseIterator {
   Status Initialize(IteratorContext* ctx) override {
     mutex_lock l(*mu_);
     if (num_parallel_calls_->value == kAutoTune) {
-      // TODO(jsimsa): Surface the number of threads used by `ctx->runner()` and
-      // use it here for the default.
-      num_parallel_calls_->value = port::NumSchedulableCPUs();
+      num_parallel_calls_->value = ctx->runner_threadpool_size();
```
[the remark this answers is not in the document]
Comment (feihugis) [reply tensorflow/core/kernels/data/parallel_map_iterator.cc]: Change it!

[the remark this answers is not in the document]
Comment (feihugis) [reply tensorflow/core/kernels/data/parallel_map_iterator.cc]: Found another one in `numa_map_and_batch_dataset_op.cc`.

**Comment by feihugis** [comment]
@jsimsa Add the change to [map_and_batch_dataset_op.cc](https://github.com/tensorflow/tensorflow/pull/23233/commits/634005cb53d2a136c04885ffa6cbf5323649fe32). Also, found another one in [numa_map_and_batch_dataset_op.cc](https://github.com/tensorflow/tensorflow/pull/23233/commits/0ff931a5965241ffdb1551493617b6d9592976e8).

Besides, change the maximum input for [model::MakeParameter::parallelism](https://github.com/tensorflow/tensorflow/pull/23233/commits/689d9974b95a764759be2028cff36735fde3001f). Is my understanding right?

**Review by jsimsa — APPROVED** [review]

**Pull request merged**